### ISSUE TITLE
messages sit in queue until GKE pod with subscriber gets reset

### ISSUE THREAD

**Issue by stephenplusplus** [googleapis/nodejs-pubsub#11]
_From @ShahNewazKhan on October 1, 2017 9:3_

#### Environment details

  - OS: Debian GNU/Linux 8.9 (jessie) [K8s pod based on dockerfile gcr.io/google_appengine/base] 
  - Node.js version: 6.11.3
  - npm version:  5.4.2
  - google-cloud/pubsub version: 0.14.2

#### Steps to reproduce

  1. Spin up nodejs pubsub publisher to topic1 in GKE pod 1
  2. Spin up nodejs pubsub subscriber to subscription to topic1 in GKE pod 2
  3. Publish messages to topic1 

I am facing an intermittent issue where pubsub messages are sitting in the queue and not being delivered to the subscriber in GKE pod 2. Only when I delete the GKE pod 2 subscriber and restart the pod does the message get delivered.  

_Copied from original issue: GoogleCloudPlatform/google-cloud-node#2640_

**Comment by stephenplusplus** [comment]
_From @callmehiphop on October 1, 2017 15:50_

We've seen a number of reports of messages not being delivered in k8s, I believe this issue is being investigated internally, although I do not know the status. @lukesneeringer have we heard any news in regards to this?

**Comment by stephenplusplus** [comment]
_From @eyalse on October 1, 2017 22:33_

I'm suffering from the same issue at the moment :(

**Comment by stephenplusplus** [comment]
_From @ApeNox on October 2, 2017 8:5_

Suffering the same issue too, please provide a fix as these are production used tools.

**Comment by stephenplusplus** [comment]
_From @eyalse on October 2, 2017 9:26_

@callmehiphop (@lukesneeringer) hey any update? as mentioned these tools (k8s and pubsub) are used in production.

**Comment by stephenplusplus** [comment]
_From @callmehiphop on October 2, 2017 14:54_

I don't have any official updates, but a new patch release was made this morning that *might* resolve the issues you're seeing.

**Comment by stephenplusplus** [comment]
_From @ShahNewazKhan on October 2, 2017 20:25_

@callmehiphop I have done some preliminary testing with the google-cloud/pubsub patch version: 0.14.3 release this morning and it looks promising so far

I have not been able to reproduce the issue yet however will need to run full end to end tests to confirm

**Comment by stephenplusplus** [comment]
_From @callmehiphop on October 2, 2017 20:34_

@ShahNewazKhan that's great, please keep us posted! 😃

**Comment by stephenplusplus** [comment]
_From @ShahNewazKhan on October 3, 2017 1:42_

@callmehiphop I have been able to replicate the issue with google-cloud/pubsub patch 0.14.3 in a slightly different use case.

Environment details

    OS: Debian GNU/Linux 8.9 (jessie) [K8s pod based on dockerfile gcr.io/google_appengine/base]
    Node.js version: 6.11.3
    npm version: 5.4.2
    google-cloud/pubsub version: 0.14.3

Steps to reproduce

1.     Spin up nodejs pubsub publisher to topic1 in GKE pod 1
2.     Spin up nodejs pubsub subscriber to subscription to topic1 in GKE pod 2
3.     Reset GKE pod 1 [pubsub publisher app]
4.     Publish messages to topic1

At this point the message remains stuck in the pubsub queue until I reset the GKE pod 2 [pubsub subscriber app]

**Comment by stephenplusplus** [comment]
_From @ShahNewazKhan on October 10, 2017 21:41_

Just checking in for updates on this issue.

**Comment by stephenplusplus** [comment]
_From @callmehiphop on October 11, 2017 18:27_

@ShahNewazKhan We believe this is a  GKE issue and because of that I can't comment on if its being worked on and when it will be fixed. I'm really sorry for the inconvenience.

**Comment by stephenplusplus** [comment]
_From @ehacke on October 16, 2017 1:25_

We may be having similar issues, not sure. @ShahNewazKhan what version of GKE are you on?

**Comment by stephenplusplus** [comment]
_From @ShahNewazKhan on October 16, 2017 2:10_

@ehacke

GKE: 1.6.10-gke.1
Kubernetes: 1.5.6

**Comment by stephenplusplus** [comment]
_From @kir-titievsky on October 26, 2017 18:23_

Question for those who'd reported this: is there any chance you had no messages published or delivered for 10 minutes or longer before you started publishing and accumulating them in the backlog?

**Comment by stephenplusplus** [comment]
_From @ShahNewazKhan on October 26, 2017 19:2_

@kir-titievsky I can confirm that the published messages sit in the subscription queue only when the publisher has been inactive longer than 10 minutes.

**Comment by stephenplusplus** [comment]
_From @kir-titievsky on October 26, 2017 19:47_

Thanks @ShahNewazKhan .  My guess here is this: by default, GCE suspends inactive connections after 10 minutes [1].  Since Pub/Sub relies on a persistent streamingPull connection, this connection would get suspended if no messages flow for 10 minutes.  This condition was not properly detected by Pub/Sub. This was fixed as of 2017-10-20 by shutting down affected streamingPull connections. The server-initiated shutdown should now trigger the client library to rebuild the connection.  

Can those of you affected check if the issue persists?


[1] https://cloud.google.com/compute/docs/troubleshooting#communicatewithinternet

**Comment by stephenplusplus** [comment]
_From @ShahNewazKhan on October 31, 2017 22:56_

@kir-titievsky Can you clarify what you mean by 'server-initiated shutdown'. Does this mean that the inactive Pub/Sub streamingPull connections are now being shutdown instead of being suspended by GCE?

I have noticed messages sitting in the queue intermittently still, do I have to update the Pub/Sub client to a latest version to handle the streamingPull connection rebuilds?

Thanks in advance!

**Comment by stephenplusplus** [comment]
I'm marking this as `blocked`, since it sounds like GKE is the party responsible for any progress on this. @callmehiphop does this sound right?

**Comment by stephenplusplus** [comment]
_From @callmehiphop on November 27, 2017 15:37_

@stephenplusplus I believe it does!

**Comment by kir-titievsky** [comment]
@ShahNewazKhan: Pub/Sub servers now close streamingPull connections regularly, with a timeout shorter than GCE's 10 minute limit. This allow the client library to quickly rebuilt the connections making sure that none are stuck in a suspended state.

**Comment by stephenplusplus** [comment]
Please let us know if there are still any issues. For now, this sounds like it's resolved.

**Comment by thomas-hilaire** [comment]
Hello,

I continue to encounter this issue regularly. As you my pod stops to receive any message after an inactive period, as soon as I restart the pod all messages are well delivered to this new pod instance.

Do I need to update anything to get the fix explained by @kir-titievsky? How can I investigate about this issue?

Thanks!

**Environment details:**
- google-cloud/pubsub: 0.15.0
- node: 8.4
 ```
$ kubectl version

Client Version: version.Info{Major:"1", Minor:"7", GitVersion:"v1.7.0", GitCommit:"d3ada0119e776222f11ec7945e6d860061339aad", GitTreeState:"clean", BuildDate:"2017-06-29T23:15:59Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}

Server Version: version.Info{Major:"1", Minor:"8+", GitVersion:"v1.8.5-gke.0", GitCommit:"2c2a807131fa8708abc92f3513fe167126c8cce5", GitTreeState:"clean", BuildDate:"2017-12-19T20:05:45Z", GoVersion:"go1.8.3b4", Compiler:"gc", Platform:"linux/amd64"}
```

**Comment by janisto** [comment]
We have experienced the same issue in the past few weeks.

A bit annoying to restart the pods daily.

**Environment details:**
-    google-cloud/pubsub: 0.13.1
-    node: 6.11

**Comment by stephenplusplus** [comment]
Re-opening in case @kir-titievsky or @callmehiphop know anything.

**Comment by janisto** [comment]
I think this was caused by grpc 1.8.4. I added grpc 1.7.3 as dependency and so far everything seems to be working fine.

**Comment by jonparrott** [comment]
Related: https://github.com/GoogleCloudPlatform/google-cloud-python/issues/4737

**Comment by domparry** [comment]
I have also solved this issue by adding grpc 1.7.3 as a dependency.

**Comment by callmehiphop** [comment]
@AlexandreAWE what version of the PubSub client are you experiencing these issues with?

**Comment by AlexandreAWE** [comment]
@callmehiphop i'm using v0.18.0. I have fixed this issue by looping every 15 minutes and restarting the subscription

**Comment by barrettc** [comment]
I'm also seeing this behavior with the latest version of google pubusb as of this comment - 0.22.2. We have two apps running in GKE communicating with each other via pubsub and the subscribers just stop receiving messages until the pods are restarted. At this point I guess I'm looking at looping and restarting the subscription every 15 minutes as described above but this feels very hacky.

**Comment by dinvlad** [comment]
I'd like to chime in and point out that we're experiencing a very similar issue, where streaming pull connections stop "restarting every 10 min" after a few times (typically in about 30 min). That seems to be correlated with modAcks not working as expected (i.e. all acks are indicated as "expired" after streaming pulls stop). This is described in https://github.com/googleapis/nodejs-pubsub/issues/314#issuecomment-461310978

**Comment by sduskis** [comment]
@plamut, this is a similar problem to the one you're seeing on Python.

**Comment by camsjams** [comment]
@AlexandreAWE @barrettc 
Just for my sake and perhaps other lurkers, how are you restarting your subscriptions?

Are you doing something straightforward like `close` and then `open` within a `setInterval` like so:
```javascript
// assuming topic
let topic = pubsub.topic('dogs');

let subscription = subscriptionTopic.subscription('my-subscription');
setInterval(async () => {
    await subscription.close();
    subscription.open();
}, 15 * 60 * 1000);
```

I would love to hear about your experiences 6 months later and if this has been reliable enough.

Thanks!

**Comment by barrettc** [comment]
@camsjams For our particular issue, it turns out we had a code path in which messages were not being properly acknowledged and the buildup of those messages created the perception that the subscriber had stopped working. In other words, we had `stupid programmer error`. I had forgotten about the comment I wrote back in December searching for answers but ultimately we did not have the problem described here in this issue.
